tests/lang/s02_core/s99_coroutines/t06_cancel_coroutine.nim

```diff
@@ -0,0 +1,28 @@
+
+## A coroutine can be cancelled, by raising an exception and letting it
+## propagate to the edge of the coroutine. Upon cancellation, the
+## ``resume`` call that started execution changes the state of the instance
+## to "aborted" and returns the instance.
+
+proc coro(cancel: bool) {.coroutine.}
+  if cancel:
+    raise CatchableError.newException("cancel")
```

Comment:
@zerbina did you consider/think of an alternative approach to cancellation, I'm trying to sort out how we might do:

1. coroutine: signals error
2. caller: recovers and resume
3. coroutine: resumes post-recovery

I imagine that'd be pretty key for enabling something like effect handling.

It could also be that `raise` is a coroutine entirely giving up (irrecoverable) vs "something else" where a coroutine needs input/something to continue or it's cancelled ('graceful' cancellation). This distinction might be entirely off base.

Reply:
> did you consider/think of an alternative approach to cancellation

I did consider an explicit `cancel`, but the way I imagined it seemed redundant with raising an exception, so I left it out.

So, with the current specification, one can do two things:
* raise, and not handle, an exception from within the coroutine (terminal; coroutine instance cannot be resumed afterwards)
* suspend and expect the caller to perform some action (like changing some coroutine instance state)

I think what you're saying, please correct me if I'm misunderstanding, is that coroutines should provide a built-in middle-ground, where the coroutine instance can signal that it can be resumed, but some action must be taken first; not doing so would raise a `CoroutineError`. How this could look like, I'm not sure yet.

---

Aside: whichever solution we pick, I think an escaping exception should abort the coroutine. My reasoning is that I think the programmer and compiler should be able reason about control-flow within a coroutine the same as they do everywhere else (minus the suspending, of course, but that should ideally be more or less opaque).

Put differently, I think that it should be possible to turn a normal procedure into a coroutine without having to change anything else, and the body continues to behave exactly the same.

If a coroutine can continue after a `raise` (or a normal call that raised), then that's no longer the above no longer holds, and whether `raise` is a terminator for structured control-flow becomes context and run-time dependent.

In short, I think `raise` should be "entirely giving up".

Reply:
> So, with the current specification, one can do two things:
> 
> * raise, and not handle, an exception from within the coroutine (terminal; coroutine instance cannot be resumed afterwards)
> * suspend and expect the caller to perform some action (like changing some coroutine instance state)
> 
> I think what you're saying, please correct me if I'm misunderstanding, is that coroutines should provide a built-in middle-ground, where the coroutine instance can signal that it can be resumed, but some action must be taken first; not doing so would raise a `CoroutineError`. How this could look like, I'm not sure yet.

Yup, I think a middle ground makes sense, to share my motivation for having a middle ground, I want to see if we can drop the `status` field, ideally more, and I believe the middle ground will be required in fulfilling that.

I'll pick up dropping the `status` field below the section break.

> Aside: whichever solution we pick, I think an escaping exception should abort the coroutine. My reasoning is that I think the programmer and compiler should be able reason about control-flow within a coroutine the same as they do everywhere else (minus the suspending, of course, but that should ideally be more or less opaque).

I also agree that raising exceptions, and their impact on control flow, should remain as expected (hah!) and therefore predictable based on existing reasoning/intuition.

> Put differently, I think that it should be possible to turn a normal procedure into a coroutine without having to change anything else, and the body continues to behave exactly the same.

This is a good property, because any background CPS transform, or the reverse, should be equivalent -- otherwise we've got bigger problems. 😬

> If a coroutine can continue after a `raise` (or a normal call that raised), then that's no longer the above no longer holds, and whether `raise` is a terminator for structured control-flow becomes context and run-time dependent.
> 
> In short, I think `raise` should be "entirely giving up".

Yup, on the same page; `raise` is an 'abort', instead of a 'fail/error'.

---

#### Dropping the `status` Field

I'd like to drop the `status` field, for possibly the same reasons @disruptek suggested removing it, mine are:

- it reduces the amount of mutable state
- the state of a coroutine is a derivation of it's internal control flow, so the field is really a cache than the reality
- it invites the temptation of fiddling with it
- smaller coroutine instances; moar fast!

Instead, at least at a low level, a coroutine should yields those state values (statically knowable within a transformed coroutine definition body). For completion, where completion might have a non-void return value a `finishedWithResult` state that informs the receiver that a 'result' field both exists and it contains a value of type `T` for a `Coroutine[T]`. This path might allow us to shrink `CoroutineBase` of other fields as well.

Otherwise, where state needs to be queried later, that burden can be carried by any runner, out of band, or by extended/more stateful coroutine instance types.